rs/nns/governance/src/proposals/add_or_remove_node_provider_tests.rs

```diff
@@ -0,0 +1,314 @@
+    let valid_proposal =
+        ValidAddOrRemoveNodeProvider::try_from(proposal).expect("Should create valid proposal");
+
+    let result = valid_proposal.validate(&node_providers);
+    assert!(
+        result.is_ok(),
+        "Expected validation to succeed for new node provider"
+    );
+}
+
+#[test]
+fn test_validate_add_existing_node_provider_fails() {
+    let node_providers = create_test_node_providers();
+    let proposal = AddOrRemoveNodeProvider {
+        change: Some(Change::ToAdd(create_test_node_provider(1))), // Already exists
+    };
+    let valid_proposal =
+        ValidAddOrRemoveNodeProvider::try_from(proposal).expect("Should create valid proposal");
+
+    let result = valid_proposal.validate(&node_providers);
+    assert_matches!(
+        result,
+        Err(error) if error.error_type == ErrorType::InvalidProposal as i32
+            && error.error_message.contains("cannot add already existing Node Provider")
+    );
+}
+
+#[test]
+fn test_validate_remove_existing_node_provider_success() {
+    let node_providers = create_test_node_providers();
+    let proposal = AddOrRemoveNodeProvider {
+        change: Some(Change::ToRemove(create_test_node_provider(1))),
+    };
+    let valid_proposal =
+        ValidAddOrRemoveNodeProvider::try_from(proposal).expect("Should create valid proposal");
+
+    let result = valid_proposal.validate(&node_providers);
+    assert!(
+        result.is_ok(),
+        "Expected validation to succeed for existing node provider"
+    );
+}
+
+#[test]
+fn test_validate_remove_non_existing_node_provider_fails() {
+    let node_providers = create_test_node_providers();
+    let proposal = AddOrRemoveNodeProvider {
+        change: Some(Change::ToRemove(create_test_node_provider(999))), // Doesn't exist
+    };
+    let valid_proposal =
+        ValidAddOrRemoveNodeProvider::try_from(proposal).expect("Should create valid proposal");
+
+    let result = valid_proposal.validate(&node_providers);
+    assert_matches!(
+        result,
+        Err(error) if error.error_type == ErrorType::InvalidProposal as i32
+            && error.error_message.contains("must target an existing Node Provider")
+    );
+}
+
+#[test]
+fn test_execute_add_node_provider_success() {
+    let mut node_providers = create_test_node_providers();
+    let new_provider = create_test_node_provider(3);
+    let proposal = AddOrRemoveNodeProvider {
+        change: Some(Change::ToAdd(new_provider.clone())),
+    };
+    let valid_proposal =
+        ValidAddOrRemoveNodeProvider::try_from(proposal).expect("Should create valid proposal");
+
+    let result = valid_proposal.execute(&mut node_providers);
+    assert!(result.is_ok(), "Expected execution to succeed");
+    assert_eq!(
+        node_providers.len(),
+        3,
+        "Expected 3 node providers after adding"
+    );
+    assert!(
+        node_providers.iter().any(|np| np.id == new_provider.id),
+        "Expected new node provider to be in the list"
+    );
```

Comment:
This is a bit weak. We should assert that the elements are np1, np2, and np3, not just that there are three elements, and at least one of the is np3. If order doesn't matter, convert from Vec to Set, or sort.